### ISSUE TITLE
let_underscore_untyped false positive with derive(Pod) on struct

### ISSUE THREAD

**Issue by rkevingibson** [rust-lang/rust-clippy#10723]
### Summary

When using the bytemuck::Pod trait (and corresponding derive macro), the lint points to a struct definition as in violation, despite having no `let` keyword in the source at all. It points to the name of the struct as the violation, and goes away if I remove the Pod trait from the list of derived traits.

### Lint Name

let_underscore_untyped

### Reproducer

I tried this code in a new blank project with the following cargo.toml dependency:
```
bytemuck = {version= "1.12.3", features= ["derive"]}
```

```rust
#![deny(clippy::all)]
#![deny(clippy::pedantic)]

use bytemuck::{Pod, Zeroable};

#[repr(C)]
#[derive(Debug, Clone, Copy, Default, Pod, Zeroable, PartialEq, PartialOrd)]
struct Foo {
    pub x: f32,
}

fn main() {
    println!("Hello, world!");
}
```

I saw this happen:

```
PS C:\Code\clippy_repro> cargo clippy
    Checking clippy_repro v0.1.0 (C:\Code\clippy_repro)
error: non-binding `let` without a type annotation
 --> src\main.rs:8:8
  |
8 | struct Foo {
  |        ^^^
  |
  = help: consider adding a type annotation or removing the `let` keyword
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#let_underscore_untyped
note: the lint level is defined here
 --> src\main.rs:2:9
  |
2 | #![deny(clippy::pedantic)]
  |         ^^^^^^^^^^^^^^^^
  = note: `#[deny(clippy::let_underscore_untyped)]` implied by `#[deny(clippy::pedantic)]`

error: could not compile `clippy_repro` due to previous error
```

I expected to see this happen:
There aren't any let keywords in the source, and as far as I can tell there's nothing in the bytemuck procedural macro that should trigger this lint.

### Version

```text
rustc 1.69.0 (84c898d65 2023-04-16)
binary: rustc
commit-hash: 84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc
commit-date: 2023-04-16
host: x86_64-pc-windows-msvc
release: 1.69.0
LLVM version: 15.0.7
```
